[Mellanox] feed module info to hw-management

jianyuewu opened the pull request:
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it
At 40°C ambient temperature with current FW+SW, some modules have >7.6% probability of reaching 75°C, which triggers false temperature warnings.
This PR implements vendor-specific temperature threshold support to eliminate false warnings while maintaining accurate temperature telemetry for monitoring purposes.

#### How I did it
Implemented new API for vendor-specific temperature offset adjustments:
1. **New  API**:
   - Add get_vendor_info() API with caching support.

2. **Smart Module Detection**:
   - Cache vendor information (Manufacturer + Part Number) for each module.
   - Skip redundant vendor info updates when the same module is replugged.

#### How to verify it
1. Plug in optical module -> Verify vendor info sent to HW-MGMT.
2. Unplug and replug same module -> Verify no redundant vendor info update.
3. Replace with different module -> Verify new vendor info sent.

#### Which release branch to backport (provide reason below if selected)
- [x] 202412

#### Tested branch (Please provide the tested image version)
202412

#### A picture of a cute animal (not mandatory but encouraged)
```bash
    /\_/\  
   ( o.o ) 
    > ^ <
   /|   |\
  (_|   |_)
   Cool Cat~
```

jianyuewu committed:
Support vendor-specific temperature adjustments

On first detection or module replacement, if the serial number (SN) has changed,
call vendor_data_set_module() with the manufacturer (MFG) and part number (PN)
to send the vendor info to hw-management.

Sample output like:
NOTICE pmon#thermalctld: Module 0 vendor info updated \
- manufacturer: NVIDIA part_number: MCP4Y10-N001

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

dgsudharsan committed:
Integrate HW-MGMT 7.0040.4036 Changes

Junchao-Mellanox commented on platform/mellanox/hw-management/0002-Remove-unused-non-upstream-kernel-modules-from-load.patch:
No actual changes?

jianyuewu replied:
This line is deleted in hw-management.hw-management.service: Wants=hw-management-sync.service

Junchao-Mellanox commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -981,24 +997,88 @@ def reinit_if_sn_changed(self):
+            try:
+                manufacturer = api.xcvr_eeprom.read(consts.VENDOR_NAME_FIELD)
+                part_number = api.xcvr_eeprom.read(consts.VENDOR_PART_NO_FIELD)
+                logger.log_debug(f"SFP {display_idx} vendor info read: manufacturer='{manufacturer}', part_number='{part_number}'")
```
Please check existing log in sfp.py to make sure if it print log with sdk_index or sdk_index + 1.

jianyuewu replied:
Yes, confirmed it will print with sdk_index+1, align with "show plat temp" print's module index.

Junchao-Mellanox commented on platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:
```diff
@@ -981,24 +997,88 @@ def reinit_if_sn_changed(self):
         """
         try:
+            sn_changed = self.reinit_if_sn_changed()
+            if sn_changed:
```
would it retry if sn_changed is False?

jianyuewu replied:
Ah, here didn't retry, when it is False. Now corrected, thanks.

jianyuewu replied:
retry_read_vendor is added, so can be easily managed in class.

jianyuewu committed:
Retry get vendor info when SFP serial change

Fix vendor-info retry to run up to 5 times in get_temperature_info on module change

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

jianyuewu commented:
Just squash the related commits, for easier review.

r12f commented:
/azp run

r12f commented:
Need to merge this PR first

https://github.com/Azure/sonic-buildimage-msft/pull/1911

jianyuewu commented:
@r12f @prgeor Could you help review this PR for Innolight feature? Thank you~

r12f commented:
Submodule merged. Merging this one.
